File: SOURCE_DOCSUITE/VecompSoftware.DocSuiteWeb.Gui.WebComponent/Grid/BindGrid.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Security.Principal;
using System.Threading;
using System.Web;
using System.Web.UI;
using Telerik.Web.UI;
using VecompSoftware.DocSuiteWeb.Data.NHibernate.Finder;
using VecompSoftware.Helpers;
using VecompSoftware.Services.Logging;


namespace VecompSoftware.DocSuiteWeb.Gui.WebComponent.Grid
{
    public class BindGrid : BaseGrid
    {
        #region Fields
        protected bool IsGrouping = false;
        protected bool IsPaging = false;
        private IFinder _finder;
        #endregion

        #region Properties
        /// <summary>
        /// Finder sulla quale è basata la griglia.
        /// </summary>
        public IFinder Finder
        {
            get
            {
                if (_finder == null)
                    _finder = GetFinderFromSession();
                return _finder;
            }
            set
            {
                _finder = value;
                PageSize = _finder.PageSize;
                if (_finder != null)
                    _finder.EnableTableJoin = true;
                PersistFinder();
            }
        }

        /// <summary>
        /// PageIndex personalizzato
        /// </summary>
        public override int CustomPageIndex
        {
            get { return Finder.CustomPageIndex; }
            set
            {
                if (Finder != null)
                {
                    Finder.CustomPageIndex = value;
                    PersistFinder();
                }
            }
        }

        #endregion

        #region Events
        public EventHandler CustomPageIndexChanged;
        #endregion

        #region Delegate
        /// <summary>
        /// Delego il chiamante a specificare la chiamata della funzione DataBindFinder
        /// </summary>
        public Action DelegateDataBindingFinder { get; set; }
        /// <summary>
        /// Delego il chiamante a specificare la chiamata della funzione FinderExecuteDoSearchHeader
        /// </summary>
        public Func<ICollection> DelegateDoSearchHeader { get; set; }

        public Action<GridSortCommandEventArgs> DelegateCreateSortingExpression { get; set; }

        #endregion

        #region Constructor
        public BindGrid() : base()
        { }
        #endregion

        #region Methods

        /// <summary>
        /// Logga le informazioni del finder
        /// </summary>
        /// <returns></returns>
        private string GetFinderSessionIdentifier()
        {
            return String.Format("{0}.{1}.IFinder", Page.GetType().BaseType.FullName, ClientID);
        }
        
        /// <summary>
        /// Legge il finder non templatizzato dalla sessione
        /// </summary>
        /// <returns></returns>
        private IFinder GetFinderFromSession()
        {
            if (HttpContext.Current == null)
                return null;
            string identifier = GetFinderSessionIdentifier();
            return (IFinder)HttpContext.Current.Session[identifier];
        }

        /// <summary>
        /// Memorizza il finder nella sessione. 
        /// Viene usato il finder non templatizzato
        /// </summary>
        public void PersistFinder()
        {
            if (HttpContext.Current == null)
                return;
            string identifier = GetFinderSessionIdentifier();
            HttpContext.Current.Session[identifier] = Finder;
            _finder = null;
            VirtualItemCount = 0; //In questo modo viene rifatto il conteggio dei record ai fini della paginazione.
        }

        /// <summary>
        /// Pulisce la sessione dal finder 
        /// </summary>
        public void DiscardFinder()
        {
            string identifier = GetFinderSessionIdentifier();
            HttpContext.Current.Session[identifier] = null;
            _finder = null;
        }

        /// <summary>
        /// Pulisce la sessione dal finder e ne impostata una nuovo
        /// </summary>
        /// <param name="finderToReplace"></param>
        public void ReplaceFinder(IFinder finderToReplace)
        {
            if(finderToReplace == null)
                DiscardFinder();
            else
                Finder = finderToReplace;
        }

        /// <summary>
        /// Mostra a video il messaggio dell'eccezione
        /// </summary>
        /// <param name="ex"></param>
        private void PrintSearchException(Exception ex)
        {
            string message = ex.Message;

            if(ex.InnerException != null) // Concateno sia l'errore che la inner exception
                message = string.Format("{0}{1}{2}", message, Environment.NewLine, ex.InnerException.Message);
            
            RadAjaxManager.GetCurrent(Page).Alert(string.Format("Errore in ricerca a causa di: {0}",StringHelper.ReplaceAlert(message, true)));
        }

        /// <summary>
        /// Questo metodo mi permette di isolare la gestione della ricerca e di settare il finder nella property.
        /// Questo metodo viene delegato allo sviluppatore. E' necessario implementarlo per eseguire il DoSearchHeader.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <typeparam name="THeader"></typeparam>
        /// <returns></returns>
        public ICollection<THeader> FinderExecuteDoSearchHeader<T, THeader>()
        {
            IFinder<T, THeader> finderToExecute = null;
            try
            {
                finderToExecute = Finder as IFinder<T, THeader>;
                if (finderToExecute != null)
                {
                    if (ImpersonateCurrentUser)
                    {
                        return ExecuteSearchWithImpersonation<ICollection<THeader>>(finderToExecute.DoSearchHeader);
                    }
                    else
                    {
                        return finderToExecute.DoSearchHeader();
                    }                    
                }                    
                return new List<THeader>();
            }
            catch (Exception ex)
            {
                PrintSearchException(ex);
                FileLogger.Warn(LogName.FileLog, String.Format("Errore nell'uso del finder [{0}]", finderToExecute.GetType()), ex);
                return null;
            }
        }

        /// <summary>
        /// Dato un finder, esegue la conta delle righe
        /// </summary>
        /// <param name="finderToExecute"></param>
        /// <returns></returns>        
        private int FinderExecuteCount(IFinder finderToExecute)
        {
            try
            {
                if (finderToExecute != null)
                {
                    Finder = finderToExecute;
                    if (ImpersonateCurrentUser)
                    {
                        return ExecuteSearchWithImpersonation<int>(finderToExecute.Count);
                    }
                    else
                    {
                        return finderToExecute.Count();
                    }                    
                }
                return 0;
            }
            catch (Exception ex)
            {
                PrintSearchException(ex);
                return default(int);
            }
        }

        private THeader ExecuteSearchWithImpersonation<THeader>(Func<THeader> func)
        {
            WindowsIdentity wi = (WindowsIdentity)HttpContext.Current.User.Identity;
            using (WindowsImpersonationContext wic = wi.Impersonate())
            using (ExecutionContext.SuppressFlow())
            {
                return func();
            }
        }

        /// <summary>
        /// Esegue il sort automatico della griglia in base alle espressioni di sort memorizzate
        /// </summary>
        /// <param name="e"></param>
        protected override void OnInit(EventArgs e)
        {
            base.OnInit(e);
            base.NeedDataSource += NeedDataSourceBinding;
        }

        /// <summary>
        /// Esegue i comandi presenti sulla griglia.
        /// 
        /// Filter
        /// CustomChangePage
        /// Export
        /// ExportFull
        /// Page
        /// ClearFilter
        /// 
        /// </summary>
        /// <param name="e"></param>
        protected override void OnItemCommand(GridCommandEventArgs e)
        {
            switch(e.CommandName)
            {
                case "Filter" :
                    {
                        CustomPageIndex = 0;
                        FinderFiltering(e);
                        DelegateDataBindingFinder();
                        break;
                    }
                case "CustomChangePage":
                    {
                        int newPageIndex;
                        if (int.TryParse(((RadTextBox)e.Item.FindControl("txtCurrentPage")).Text, out newPageIndex))
                        {
                            CustomPageIndex = newPageIndex - 1;
                            DelegateDataBindingFinder();
                        }
                        break;
                    }
                case "Export" : 
                    {
                        DoExport(e);
                        break;
                    }
                case "ExportFull" :
                    {
                        // TODO: riscrivere la procedura di export. Utilizzare una copia del finder per estrarre i dati da mettere sull'esportazione e mappare le colonne già presenti nella griglia.

                        // Il finder viene chiamato 2 volte:
                        // 1) la prima chiamata serve per riempire tutta la griglia e poi lanciare l'export.
                        // 2) La seconda chiamata invece serve per resettare i dati in gliglia e mettere solo quelli visualizzabili.
                        int oldPageSize = Finder.PageSize, oldPageIndex = Finder.CustomPageIndex,
                            oldGridPageSize = this.PageSize, oldGridPageCount = this.PageCount;
                        // 1)
                        base.PageSize = this.PageCount * this.PageSize;
                        base.CustomPageIndex = 0;
                        Finder.PageSize = this.PageCount * this.PageSize;                        
                        Finder.CustomPageIndex = 0;
                        DelegateDataBindingFinder();
                        GridColumnCollection gridColumns = this.Columns;
                        GridDataItemCollection gridItems = this.Items;

                        // 2)
                        this.PageSize = oldGridPageSize;
                        Finder.PageSize = oldPageSize;
                        Finder.CustomPageIndex = oldPageIndex;
                        DelegateDataBindingFinder();

                        // esporto le colonne e gli elementi presenti nella griglia completa.
                        DoExport(e, gridColumns, gridItems);                        
                        break;
                    }
                case "Page":
                    {
                        DoPaging(e);
                        break;
                    }
                case "ClearFilters":
                    {
                        // TODO: rifare il metodo di pulizia filtri. 
                        /*
                        FilterHelper.RemoveFilterClientState(Page.Session, CommonUtil.GetInstance().AppTempPath, CustomPageIndex + 1);
                        */
                        
                        Finder.FilterExpressionsClear();                        
                        PersistFinder();
                        DelegateDataBindingFinder();                        
                        break;
                    }
                default:
                    {
                        base.OnItemCommand(e);
                        break;
                    }
            }
        }

        /// <summary>
        /// Esegue la paginazione della griglia leggendo i parametri di dimensione
        /// </summary>
        /// <param name="e"></param>
        protected override void OnSortCommand(GridSortCommandEventArgs e)
        {
            base.OnSortCommand(e);
            DelegateCreateSortingExpression(e);
        }

        protected override void OnItemCreated(GridItemEventArgs e)
        {
            base.OnItemCreated(e);
            
            if (!(e.Item is GridFilteringItem) || Finder == null)
                return;
            
            // TODO rivedere l'espressione di filtraggio. 
            /*
               If Finder.FilterExpressions Is Nothing Then
                   Exit Sub
               End If
               
               Dim filteringItem As GridFilteringItem = DirectCast(e.Item, GridFilteringItem)
               For Each columnName As String In Finder.FilterExpressions.Keys
                   Dim control As Control = FilterHelper.GetFilterControl(Of Control)(filteringItem(columnName).Controls(0))
                   FilterHelper.SetFilterValue(control, Finder.FilterExpressions(columnName).FilterValue)
               Next
             */

            // TODO: rivedere il metodo di creazione dell'oggetto            
        }

        private void BindGrid_GroupsChanging(object source, GridGroupsChangingEventArgs e)
        {
            IsGrouping = true;
            if (e.Action == GridGroupsChangingAction.Ungroup)
                IsGrouping = false;
        }

        private void FinderFiltering(GridCommandEventArgs e)
        {
            // annullo il comando
            e.Canceled = true;
            
            // TODO: rivedere la logica di applicazione dei filtri
            
            //aggiorna filtro nel ViewState
            PersistFinder();             
        }

        /// <summary>
        /// Esegue il sort di una colonna nella griglia.
        /// Reperisco dalla griglia il nome della colonna da ordinare.
        /// Creo l'espressione Lambda corretta e la inserisco nel finder e poi lo persisto nella sessione.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <typeparam name="THeader"></typeparam>
        /// <param name="e"></param>
        public void FinderSorting<T, THeader>(GridSortCommandEventArgs e)
        {            
            // TODO: verificare gli ordinamenti sui campi dell'entità Header. In questo momento l'ordinamento funziona solo se il campo è settato sull'entità padre

            IFinder<T, THeader> finderToExecute = Finder as IFinder<T, THeader>;
            finderToExecute.SortExpressionsClear();
            SortDirection sortOrder;
            if (e.NewSortOrder == GridSortOrder.None && e.OldSortOrder == GridSortOrder.Descending)
                sortOrder = SortDirection.Descending;
            else
                sortOrder = e.NewSortOrder == GridSortOrder.Ascending ? SortDirection.Ascending : SortDirection.Descending;
            
            string parameterName = e.SortExpression;
            ParameterExpression param = Expression.Parameter(typeof(T), "x");
            Expression<Func<T, object>> expr = Expression.Lambda<Func<T, object>>(Expression.Convert(Expression.Property(param, parameterName), typeof(object)), param);
            finderToExecute.SortExpressions.Add(new SortExpression<T>(){Expression = expr, Direction = sortOrder});
            Finder = finderToExecute;

            PersistFinder();
        }
        
        /// <summary>
        /// Esegue la paginazione
        /// </summary>
        /// <param name="e"></param>
        private void DoPaging(GridCommandEventArgs e)
        {
            base.OnItemCommand(e);
            
            IsPaging = true;
            switch (e.CommandArgument.ToString())
            {
                case "First":
                    CustomPageIndex = 0;
                    if (CustomPageIndexChanged != null)
                    {
                        CustomPageIndexChanged(this, new EventArgs());
                    }

                    DelegateDataBindingFinder();

                    break;
                case "Prev":
                    CustomPageIndex = CustomPageIndex - 1;
                    if (CustomPageIndexChanged != null)
                    {
                        CustomPageIndexChanged(this, new EventArgs());
                    }

                    DelegateDataBindingFinder();

                    break;
                case "Next":
                    CustomPageIndex = CustomPageIndex + 1;
                    if (CustomPageIndexChanged != null)
                    {
                        CustomPageIndexChanged(this, new EventArgs());
                    }

                    DelegateDataBindingFinder();

                    break;
                case "Last":
                    CustomPageIndex = PageCount - 1;
                    if (CustomPageIndexChanged != null)
                    {
                        CustomPageIndexChanged(this, new EventArgs());
                    }

                    DelegateDataBindingFinder();

                    break;
                default:
                    IsPaging = false;
                    break;
            }
        }

        /// <summary>
        /// Binding del data source
        /// </summary>
        /// <param name="source"></param>
        /// <param name="e"></param>
        private void NeedDataSourceBinding(object source, GridNeedDataSourceEventArgs e)
        {
            ICollection results = DelegateDoSearchHeader();
            DataSource = results;

            if (results != null)
            {
                InitializeVirtualItemCount(Finder);
                
                if (((MasterTableView.GroupByExpressions.Count > 0) || IsGrouping))
                {
                    CurrentPageIndex = 0;
                }
            }            
        }

        /// <summary>
        /// Esegue la ricerca attraverso il finder interno eseguendo anche il count dei risultati totali
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <typeparam name="THeader"></typeparam>
        public void DataBindFinder<T, THeader>()
        {
            IFinder<T, THeader> finderToExecute = Finder as IFinder<T, THeader>;
            if (finderToExecute == null)
            {
                return;
            }
            ICollection<THeader> results = FinderExecuteDoSearchHeader<T, THeader>();
            DataSource = results;
            
            if (results != null)
            {
                InitializeVirtualItemCount(Finder);
                
                if (((MasterTableView.GroupByExpressions.Count > 0) || IsGrouping))
                {
                    CurrentPageIndex = 0;
                }
                //TODO: Inserire l'ordinamento qui per la parte di esportazione
            }
            
            DataBind();
        }
        
        /// <summary>
        /// Inizializza il conteggio delle righe in griglia
        /// </summary>
        /// <param name="finderToExecute"></param>
        private void InitializeVirtualItemCount(IFinder finderToExecute)
        {
            if (VirtualItemCount == 0)
            {
                VirtualItemCount = FinderExecuteCount(finderToExecute);
            }
        }

        /// <summary>
        /// Esegue solo la ricerca attraverso il finder interno.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <typeparam name="THeader"></typeparam>
        /// <param name="finderToExecute"></param>
        /// <param name="externalCount">Count dei risultati totali</param>
        public void DataBindFinderWithExtCount<T, THeader>(IFinder<T, THeader> finderToExecute, int externalCount)
        {
            ICollection<THeader> results = FinderExecuteDoSearchHeader<T, THeader>();
            DataSource = results;

            if (results != null)
            {
                VirtualItemCount = externalCount;                
            }
            DataBind();
        }

        public void SaveSettings()
        {
            // TODO: salvare i filtri sulla griglia

            /*
            HttpCookie myCookie = new HttpCookie(ClientID + "_GridSettings");
            myCookie(ClientID + "_state") = GridSettingsPersister.SaveSettings(this);
            Page.Response.Cookies.Add(myCookie);
            */
        }

        public void LoadSettings(string settings)
        {
            // TODO: caricare i filtri sulla griglia

            /*
            HttpCookie myCookie = Page.Request.Cookies(ClientID + "_GridSettings");
            if ((myCookie != null && myCookie(ClientID + "_state") != null))
            {
                GridSettingsPersister.LoadSettings(this, myCookie(ClientID + "_state"));
            }
             */
        }
        #endregion

        /// <summary>
        /// Necessario per reimpostare la pagina precedente in "History Back".
        /// </summary>
        /// <param name="e"></param>
        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            Page.Response.Cache.SetAllowResponseInBrowserHistory(false);
            Page.Response.Cache.SetCacheability(HttpCacheability.NoCache);
            Page.Response.Cache.SetNoStore();
            Page.Response.Cache.SetValidUntilExpires(true);
        }
    }
}

File: SOURCE_DOCSUITE/VecompSoftware.DocSuiteWeb.Presentation/Grid/BindGrid.vb
Imports System.Web.UI
Imports System.Linq
Imports VecompSoftware.DocSuiteWeb.Facade
Imports VecompSoftware.DocSuiteWeb.Data
Imports Telerik.Web.UI
Imports VecompSoftware.Helpers
Imports VecompSoftware.Services.Logging
Imports System.Web
Imports VecompSoftware.WebAPIManager.Exceptions
Imports VecompSoftware.DocSuiteWeb.DTO.WebAPI
Imports VecompSoftware.DocSuiteWeb.Data.WebAPI.Finder.Fascicles

Public Class BindGrid
    Inherits BaseGrid

#Region " Fields "

    Protected IsGrouping As Boolean = False
    Protected IsPaging As Boolean = False

    Private _finder As IFinder

#End Region

#Region " Properties "

    Public Property Finder As IFinder
        Get
            If _finder Is Nothing Then
                _finder = GetFinderFromSession()
            End If
            Return _finder
        End Get
        Set(ByVal value As IFinder)
            _finder = value
            PageSize = _finder.PageSize
            If _finder IsNot Nothing Then
                ReflectionHelper.SetProperty(_finder, "EnableTableJoin", True)
            End If
            PersistFinder()
        End Set
    End Property


    Public Overrides Property CustomPageIndex As Integer
        Get
            Return Finder.CustomPageIndex
        End Get
        Set(ByVal value As Integer)
            Finder.CustomPageIndex = value
            PersistFinder()
        End Set
    End Property

    Public Property Criteria As Object
        Get
            Return Page.Session(ID + "_criteria")
        End Get
        Set(ByVal value As Object)
            Page.Session(ID + "_criteria") = value
        End Set
    End Property

    Private Property ImpersonationAction As Func(Of IFinder, Func(Of Object), Object)

    Private Property ImpersonationCounterAction As Func(Of IFinder, Func(Of Integer), Integer)

#End Region

#Region " Constructor "

    Public Sub New()
        MyBase.New()
    End Sub

#End Region

#Region " Methods "

    ' Gestione finder in sessione
    Private Function GetFinderSessionIdentifier() As String
        Return String.Format("{0}.{1}.IFinder", Page.GetType().BaseType.FullName, ClientID)
    End Function

    Private Function GetFinderFromSession() As IFinder
        If HttpContext.Current Is Nothing Then
            Return Nothing
        End If
        Dim identifier As String = GetFinderSessionIdentifier()
        Return DirectCast(HttpContext.Current.Session.Item(identifier), IFinder)
    End Function

    Public Sub PersistFinder()
        If HttpContext.Current Is Nothing Then
            Return
        End If
        Dim identifier As String = GetFinderSessionIdentifier()
        HttpContext.Current.Session.Item(identifier) = Finder
        _finder = Nothing
        VirtualItemCount = 0 ' In questo modo viene rifatto il conteggio dei record ai fini della paginazione.
    End Sub

    Public Sub DiscardFinder()
        Dim identifier As String = GetFinderSessionIdentifier()
        HttpContext.Current.Session.Item(identifier) = Nothing
        _finder = Nothing
    End Sub

    Public Sub ReplaceFinder(finderToReplace As IFinder)
        If finderToReplace Is Nothing Then
            DiscardFinder()
        Else
            Finder = finderToReplace
        End If
    End Sub

    ''' <summary> Mostra a video il messaggio dell'eccezione </summary>
    Private Sub PrintSearchException(ByVal ex As Exception)
        Dim message As String = ex.Message
        If ex.InnerException IsNot Nothing Then
            message = ex.InnerException.Message
        End If
        FileLogger.Error(LogName.FileLog, message, ex)
        RadAjaxManager.GetCurrent(Page).Alert("Errore in ricerca a causa di: " & StringHelper.ReplaceAlert(message, True))
    End Sub

    Private Sub PrintWebAPIException(ByVal ex As WebAPIException(Of ICollection(Of WebAPIDto(Of Object))))
        Dim message As String = ex.Message
        If ex.InnerException IsNot Nothing Then
            message = ex.InnerException.Message
        End If
        FileLogger.Error(LogName.FileLog, message, ex)
        RadAjaxManager.GetCurrent(Page).Alert(String.Format("Non è stato possibile ricevere informazioni da {0}.", ex.TenantName))
    End Sub

    Private Function FinderExecuteDoSearchHeader(finderToExecute As Object) As Object
        Try
            Finder = finderToExecute
            If ImpersonateCurrentUser Then
                RaiseEvent NeedImpersonation(Me, New EventArgs())
                If TypeOf Finder Is FascicleFinder AndAlso finderToExecute.FromPostMethod Then
                    Return ExecuteSearchWithImpersonation(Of Object)(AddressOf finderToExecute.GetFromPostMethod)
                End If
                Return ExecuteSearchWithImpersonation(Of Object)(AddressOf finderToExecute.DoSearchHeader)
            Else
                If TypeOf Finder Is FascicleFinder AndAlso finderToExecute.FromPostMethod Then
                    Return finderToExecute.GetFromPostMethod()
                End If
                Return finderToExecute.DoSearchHeader()
            End If
        Catch ex_ea As AggregateException
            For Each ex As Exception In ex_ea.InnerExceptions
                If TypeOf ex Is WebAPIException(Of ICollection(Of WebAPIDto(Of Object))) Then
                    PrintWebAPIException(ex)
                    Return CType(ex, WebAPIException(Of ICollection(Of WebAPIDto(Of Object)))).Results
                Else
                    FileLogger.Error(LogName.FileLog, ex.Message, ex)
                End If
            Next
        Catch ex As WebAPIException(Of ICollection(Of WebAPIDto(Of Object)))
            PrintWebAPIException(ex)
            Return ex.Results
        Catch ex As Exception
            PrintSearchException(ex)
            FileLogger.Warn(LogName.FileLog, String.Format("Errore nell'uso del finder [{0}]", finderToExecute.GetType()), ex)
        End Try
        Return New ArrayList()
    End Function

    Private Function FinderExecuteCount(finderToExecute As Object) As Integer
        Try
            Finder = finderToExecute
            If ImpersonateCurrentUser Then
                RaiseEvent NeedImpersonation(Me, New EventArgs())
                Return ExecuteCounterWithImpersonation(AddressOf finderToExecute.Count)
            Else
                Return finderToExecute.Count()
            End If
        Catch ex As Exception
            PrintSearchException(ex)
        End Try
    End Function

    Public Sub SetImpersonationAction(func As Func(Of IFinder, Func(Of Object), Object))
        ImpersonationAction = func
    End Sub

    Public Sub SetImpersonationCounterAction(func As Func(Of IFinder, Func(Of Integer), Integer))
        ImpersonationCounterAction = func
    End Sub

    Private Function ExecuteSearchWithImpersonation(Of T)(func As Func(Of T)) As T
        If ImpersonationAction Is Nothing Then
            Return func()
        End If

        Return ImpersonationAction(Finder, func)
    End Function

    Private Function ExecuteCounterWithImpersonation(func As Func(Of Integer)) As Integer
        If ImpersonationCounterAction Is Nothing Then
            Return func()
        End If

        Return ImpersonationCounterAction(Finder, func)
    End Function

    ''' <summary> Esegue il sort automatico della griglia in base alle espressioni di sort memorizzate </summary>
    Protected Overrides Sub OnInit(ByVal e As EventArgs)
        MyBase.OnInit(e)
        AddHandler NeedDataSource, AddressOf NeedDataSourceDelegate
    End Sub


    Protected Overrides Sub OnItemCommand(ByVal e As GridCommandEventArgs)
        Select Case e.CommandName
            Case "Filter"
                'MasterTableView.AllowMultiColumnSorting = True
                CustomPageIndex = 0
                FinderFiltering(e)
                DataBindFinder()
                'MasterTableView.AllowMultiColumnSorting = False
                Exit Select
            Case "CustomChangePage"
                Dim newPageIndex As Integer
                If Integer.TryParse(DirectCast(e.Item.FindControl("txtCurrentPage"), RadTextBox).Text, newPageIndex) Then
                    CustomPageIndex = newPageIndex - 1
                    RaiseEvent CustomPageIndexChanged(Me, New EventArgs())
                    DataBindFinder()
                End If
            Case "Export"
                DoExport(e, Me.Columns, Me.Items)
            Case "ExportFull"
                Dim oldPageSize = Finder.PageSize
                Dim oldPageIndex = Finder.CustomPageIndex
                Dim oldGridPageSize = Me.PageSize
                Dim oldGridPageCount = Me.PageCount
                Finder.PageSize = Me.PageCount * Me.PageSize
                Finder.CustomPageIndex = 0
                Me.DataBindFinder()
                Dim gridColumns As GridColumnCollection = Me.Columns
                Dim gridItems As GridDataItemCollection = Me.Items
                Me.PageSize = oldGridPageSize
                Finder.PageSize = oldPageSize
                Finder.CustomPageIndex = oldPageIndex
                Me.DataBindFinder()
                DoExport(e, gridColumns, gridItems)
            Case "Page"
                DoPaging(e)
            Case "ClearFilters"
                FilterHelper.RemoveFilterClientState(Page.Session, CommonUtil.GetInstance().AppTempPath, CustomPageIndex + 1)
                Finder.FilterExpressions.Clear()
                PersistFinder()
                DataBindFinder()
            Case "CustomFilter"
                CustomPageIndex = 0
                Dim commands As IList(Of IFilterExpression) = DirectCast(e.CommandArgument, IList(Of IFilterExpression))
                For Each command As IFilterExpression In commands
                    Dim filterExpressions As IDictionary(Of String, IFilterExpression) = ReflectionHelper.GetProperty(Finder, "FilterExpressions")
                    FilterHelper.UpdateFilter(filterExpressions, command)
                Next
                DataBindFinder()

            Case Else
                MyBase.OnItemCommand(e)
        End Select
    End Sub

    Protected Sub NeedDataSourceDelegate(ByVal source As Object, ByVal e As GridNeedDataSourceEventArgs)
        If Not IsPaging Then
            NeedDataSourceBinding()
        End If
    End Sub

    ''' <summary> Esegue la paginazione della griglia leggendo i parametri di dimensione </summary>
    Protected Overrides Sub OnSortCommand(ByVal e As GridSortCommandEventArgs)
        MyBase.OnSortCommand(e)
        FinderSorting(e)
    End Sub

    Protected Overrides Sub OnItemCreated(ByVal e As GridItemEventArgs)
        MyBase.OnItemCreated(e)

        If Not (TypeOf e.Item Is GridFilteringItem) OrElse (Finder Is Nothing) Then
            Exit Sub
        End If

        If Finder.FilterExpressions Is Nothing Then
            Exit Sub
        End If

        Dim filteringItem As GridFilteringItem = DirectCast(e.Item, GridFilteringItem)
        For Each columnName As String In Finder.FilterExpressions.Keys
            Try
                Dim control As Control = FilterHelper.GetFilterControl(Of Control)(filteringItem(columnName).Controls(0))
                FilterHelper.SetFilterValue(control, Finder.FilterExpressions(columnName).FilterValue)
            Catch ex As Exception
                FileLogger.Error(LogName.FileLog, ex.Message, ex)
            End Try
        Next
    End Sub

    Private Sub BindGrid_GroupsChanging(ByVal source As Object, ByVal e As GridGroupsChangingEventArgs) Handles Me.GroupsChanging
        IsGrouping = True
        If e.Action = GridGroupsChangingAction.Ungroup Then
            IsGrouping = False
        End If
    End Sub

    Private Sub FinderFiltering(ByVal e As GridCommandEventArgs)
        'annullo il comando
        e.Canceled = True

        'Istanzia il builder dei filtri dopo averlo recuperato
        Dim command As Pair = DirectCast(e.CommandArgument, Pair)
        Dim filterSetter As New FilterSetter(Columns.FindByUniqueName(command.Second), Finder)
        'imposta il filtro
        Dim filterItem As GridFilteringItem = DirectCast(e.Item, GridFilteringItem)
        filterSetter.SetFilter(command, filterItem)

        'evidenzia il filtro corrente
        FilterHelper.HighlightFilterMenu(command.First, FilterMenu)

        'aggiorna filtro nel ViewState
        PersistFinder()
    End Sub

    Private Sub FinderSorting(ByVal e As GridSortCommandEventArgs)
        Finder.SortExpressions.Clear()

        Dim sortOrder As String
        If e.NewSortOrder = GridSortOrder.None AndAlso e.OldSortOrder = GridSortOrder.Descending Then
            sortOrder = "ASC"
        Else
            sortOrder = If(e.NewSortOrder = GridSortOrder.Ascending, "ASC", "DESC")
        End If
        Finder.SortExpressions.Add(e.SortExpression, sortOrder)

        PersistFinder()
    End Sub


    Private Sub AttachSortExpressionFromFinder()
        Dim dic As IDictionary(Of String, String) = Finder.SortExpressions
        If dic.Keys.Count <= 0 Then
            Exit Sub
        End If

        '' Il codice precedente inseriva gli ordinamenti n-volte
        '' Questo verifica solo se ci sono sort expression da collegare e nel caso lo fa una volta sola
        If Columns.Cast(Of GridColumn)().Any(Function(col) dic.Keys.Contains(col.SortExpression)) Then
            MasterTableView.SortExpressions.Clear()
            For Each sortExpression As KeyValuePair(Of String, String) In dic
                MasterTableView.SortExpressions.AddSortExpression(String.Format("{0} {1}", sortExpression.Key, sortExpression.Value))
            Next
        End If
    End Sub

    Private Sub DoPaging(ByVal e As GridCommandEventArgs)
        MyBase.OnItemCommand(e)

        IsPaging = True
        Select Case e.CommandArgument
            Case "First"
                CustomPageIndex = 0
                RaiseEvent CustomPageIndexChanged(Me, New EventArgs())
                DataBindFinder()

            Case "Prev"
                CustomPageIndex = CustomPageIndex - 1
                RaiseEvent CustomPageIndexChanged(Me, New EventArgs())
                DataBindFinder()

            Case "Next"
                CustomPageIndex = CustomPageIndex + 1
                RaiseEvent CustomPageIndexChanged(Me, New EventArgs())
                DataBindFinder()

            Case "Last"
                CustomPageIndex = PageCount - 1
                RaiseEvent CustomPageIndexChanged(Me, New EventArgs())
                DataBindFinder()

            Case Else
                IsPaging = False
        End Select
    End Sub

    Private Sub NeedDataSourceBinding()
        If Finder Is Nothing Then
            Exit Sub
        End If

        DataSource = FinderExecuteDoSearchHeader(Finder)
        If DataSource.Count > 0 Then
            InitializeVirtualItemCount()

            If ((MasterTableView.GroupByExpressions.Count > 0) OrElse IsGrouping) Then
                CurrentPageIndex = 0
            End If
            AttachSortExpressionFromFinder()
        End If
    End Sub

    ''' <summary> Esegue la ricerca attraverso il finder interno eseguendo anche il count dei risultati totali </summary>
    Public Sub DataBindFinder()
        DataSource = FinderExecuteDoSearchHeader(Finder)
        If DataSource.Count > 0 Then
            InitializeVirtualItemCount()

            If ((MasterTableView.GroupByExpressions.Count > 0) OrElse IsGrouping) Then
                CurrentPageIndex = 0
            End If
            AttachSortExpressionFromFinder()
        End If

        DataBind()
    End Sub

    Private Sub InitializeVirtualItemCount()
        If VirtualItemCount = 0 Then
            VirtualItemCount = FinderExecuteCount(Finder)
        End If
    End Sub

    ''' <summary> Esegue solo la ricerca attraverso il finder interno. </summary>
    ''' <param name="externalCount">Count dei risultati totali</param>
    Public Sub DataBindFinderWithExtCount(ByVal externalCount As Integer)

        DataSource = FinderExecuteDoSearchHeader(Finder)

        If DataSource.Count > 0 Then
            VirtualItemCount = externalCount
            AttachSortExpressionFromFinder()
        End If

        DataBind()
    End Sub

    Public Sub SaveSettings()
        Dim myCookie As New HttpCookie(ClientID & "_GridSettings")
        myCookie(ClientID & "_state") = GridSettingsPersister.SaveSettings(Me)
        Page.Response.Cookies.Add(myCookie)
    End Sub

    Public Sub LoadSettings(ByVal settings As String)
        Dim myCookie As HttpCookie = Page.Request.Cookies(ClientID & "_GridSettings")
        If (myCookie IsNot Nothing AndAlso myCookie(ClientID & "_state") IsNot Nothing) Then
            GridSettingsPersister.LoadSettings(Me, myCookie(ClientID & "_state"))
        End If
    End Sub

#End Region

#Region " Events "

    Private Sub BindGrid_Load(sender As Object, e As EventArgs) Handles Me.Load
        ' FG20130909: Necessario per reimpostare la pagina precedente in "History Back".
        Page.Response.Cache.SetAllowResponseInBrowserHistory(False)
        Page.Response.Cache.SetCacheability(HttpCacheability.NoCache)
        Page.Response.Cache.SetNoStore()
        Page.Response.Cache.SetValidUntilExpires(True)
    End Sub

#End Region

    Public Event CustomPageIndexChanged As EventHandler

    Public Sub OnCustomPageIndexChanged(e As EventArgs)
        RaiseEvent CustomPageIndexChanged(Me, e)
    End Sub

    Public Event NeedImpersonation As EventHandler

    Public Sub OnNeedImpersonation(e As EventArgs)
        RaiseEvent NeedImpersonation(Me, e)
    End Sub

End Class